### PR TITLE
add skip secondary authorization option for public oauth2 clients

### PR DESCRIPTION
**Pull request by denyskon**
Followup to #30790

Add an option to allow skipping the manual authorization on every request for public clients.

As this measure is not a hard requirement by the OAuth2 spec and all the major competitors of Gitea not implementing it, make it opt-out to allow for consistent user experience and simplified workflow for users who need it.

### CODE REVIEW

**Comment by denyskon** [comment]
(easier to review with disabled whitespace changes :)

File: web_src/js/features/oauth2-settings.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+export function initOAuth2SettingsDisableCheckbox() {
+  for (const e of document.querySelectorAll('.disable-setting')) e.addEventListener('change', ({target}) => {
+    if (target.checked) {
+      document.querySelector(e.getAttribute('data-target')).classList.add('disabled');
+    } else {
+      document.querySelector(e.getAttribute('data-target')).classList.remove('disabled');
+    }
```
Comment (delvh) [inline web_src/js/features/oauth2-settings.js]:
```suggestion
    document.querySelector(e.getAttribute('data-target')).classList.toggle('disabled', target.checked);
```

File: models/migrations/v1_23/v300.go [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+package v1_23 //nolint
+
+import (
+	"xorm.io/xorm"
+)
+
+// AddSkipSeconderyAuthToOAuth2ApplicationTable: add SkipSecondaryAuthorization column, setting existing rows to false
+func AddSkipSecondaryAuthColumnToOAuth2ApplicationTable(x *xorm.Engine) error {
+	type oauth2Application struct {
+		ID                         int64
```
Comment (delvh) [inline models/migrations/v1_23/v300.go]:
```suggestion
```

File: routers/web/auth/oauth.go [hunk 3]
```diff
@@ -470,9 +470,9 @@ func AuthorizeOAuth(ctx *context.Context) {
 		return
 	}
 
-	// Redirect if user already granted access and the application is confidential.
-	// I.e. always require authorization for public clients as recommended by RFC 6749 Section 10.2
-	if app.ConfidentialClient && grant != nil {
+	// Redirect if user already granted access and the application is confidential or trusted otherwise
+	// I.e. always require authorization for untrusted public clients as recommended by RFC 6749 Section 10.2
+	if (app.ConfidentialClient || app.SkipSecondaryAuthorization) && grant != nil {
```
Comment (delvh) [inline routers/web/auth/oauth.go:474]: This code reads like it should be
```suggestion
	if (app.ConfidentialClient || !app.SkipSecondaryAuthorization) && grant != nil {
```
or am I missing something?
It looks as if you say `if skip authorization then ask user to authorize themselves` right now.

Comment (denyskon) [reply routers/web/auth/oauth.go:474]: No, basically the code says:

`if access already granted AND (confidential client OR skip secondary auth) then use the existing grant to authorize the application directly`

**Review by delvh — APPROVED** [review]
I don't think I'm able to test this, so here's my approval.

File: options/locale/locale_en-US.ini [hunk 4]
```diff
@@ -911,6 +911,7 @@ create_oauth2_application_success = You have successfully created a new OAuth2 a
 update_oauth2_application_success = You have successfully updated the OAuth2 application.
 oauth2_application_name = Application Name
 oauth2_confidential_client = Confidential Client. Select for apps that keep the secret confidential, such as web apps. Do not select for native apps including desktop and mobile apps.
+oauth2_skip_secondary_authorization = Skip authorization for public client after granting access once. <b>May pose a security risk.</b>
```
Comment (techknowlogick) [inline options/locale/locale_en-US.ini]:
```suggestion
oauth2_skip_secondary_authorization = Skip authorization for public clients after granting access once. <strong>May pose a security risk.</strong>
```

**Review by techknowlogick — APPROVED** [review]

**Pull request merged**